kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ChecksumReader.java

```diff
@@ -0,0 +1,119 @@
+import static io.delta.kernel.internal.util.FileNames.isChecksumFile;
+import static io.delta.kernel.internal.util.Utils.singletonCloseableIterator;
+
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.util.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility method to load protocol and metadata from the Delta log checksum files. */
+public class ChecksumReader {
+  private static final Logger logger = LoggerFactory.getLogger(ChecksumReader.class);
+
+  /**
+   * Load the CRCInfo from the checksum file at the given version. If the checksum file is not found
+   * at the given version, it will try to find the latest checksum file that is created after the
+   * lower bound version or within the last 100 versions.
+   *
+   * @param engine the engine to use for reading the checksum file
+   * @param logPath the path to the Delta log
+   * @param targetedVersion the target version to read the checksum file from
+   * @param lowerBound the exclusive lower bound version to search for the checksum file
+   * @return Optional {@link CRCInfo} containing the protocol and metadata, and the version of the
+   *     checksum file. If the checksum file is not found, it will return an empty
+   */
+  public static Optional<CRCInfo> getCRCInfo(
+      Engine engine, Path logPath, long targetedVersion, long lowerBound) {
+    logger.info("Loading CRC file for version {} with lower bound {}", targetedVersion, lowerBound);
+    // First try to load the CRC at given version. If not found or failed to read then try to
+    // find the latest CRC file that is created after the lower bound version or within the last 100
+    // versions if no lower bound is provided.
+    Path crcFilePath = checksumFile(logPath, targetedVersion);
+    Optional<CRCInfo> crcInfoOpt = readChecksumFile(engine, crcFilePath);
+    if (crcInfoOpt.isPresent()
+        ||
+        // we don't expect any more checksum files as it is the first version
+        targetedVersion == 0) {
+      return crcInfoOpt;
+    }
+    logger.info(
+        "CRC file for version {} not found, attempt to loading version up to {}",
+        targetedVersion,
+        lowerBound);
+
+    Path lowerBoundFilePath = checksumFile(logPath, lowerBound + 1);
+    try (CloseableIterator<FileStatus> crcFiles =
+        engine.getFileSystemClient().listFrom(lowerBoundFilePath.toString())) {
+      List<FileStatus> crcFilesList = new ArrayList<>();
+      crcFiles.filter(file -> isChecksumFile(file.getPath())).forEachRemaining(crcFilesList::add);
+
+      // pick the last file which is the latest version that has the CRC file
+      if (crcFilesList.isEmpty()) {
+        logger.warn("No checksum files found in the range {} to {}", lowerBound, targetedVersion);
```

Comment:
printing the lower bound here will be misleading since this is the exclusive lowerBound.

what was the reasoning behind making lowerbound exclusive? was it worth the tiny little mistakes like this I wonder?

Reply:
I updated to make in inclusive, pervious exclusive implementation has two issues mentioned above when checkpoint came into picture.